pkg/genericapiserver/storage_factory.go

```diff
@@ -80,48 +65,132 @@ type DefaultStorageFactory struct {
 	newEtcdFn func(ns runtime.NegotiatedSerializer, storageVersion, memoryVersion unversioned.GroupVersion, etcdConfig etcdstorage.EtcdConfig) (etcdStorage storage.Interface, err error)
 }
 
+type groupResourceOverrides struct {
+	// etcdLocation contains the list of "special" locations that are used for particular GroupResources
+	// These are merged on top of the default DefaultEtcdConfig when requesting the storage.Interface for a given GroupResource
+	etcdLocation *[]string
+	// etcdPrefix contains the list of "special" prefixes for a GroupResource.  Resource=* means for the entire group
+	etcdPrefix *string
+	// serializer contains the list of "special" serializers for a GroupResource.  Resource=* means for the entire group
+	serializer *runtime.NegotiatedSerializer
+	// cohabitatingResources keeps track of which resources must be stored together.  This happens when we have multiple ways
+	// of exposing one set of concepts.  autoscaling.HPA and extensions.HPA as a for instance
+	// The order of the slice matters!  It is the priority order of lookup for finding a storage location
+	cohabitatingResources *[]unversioned.GroupResource
+}
+
 const AllResources = "*"
 
 func NewDefaultStorageFactory(defaultEtcdConfig etcdstorage.EtcdConfig, defaultSerializer runtime.NegotiatedSerializer, resourceEncodingConfig ResourceEncodingConfig, resourceConfig APIResourceConfigSource) *DefaultStorageFactory {
 	return &DefaultStorageFactory{
 		DefaultEtcdConfig:       defaultEtcdConfig,
-		EtcdLocationOverrides:   map[unversioned.GroupResource][]string{},
-		EtcdPrefixOverrides:     map[unversioned.GroupResource]string{},
+		GroupResourceOverrides:  map[unversioned.GroupResource]groupResourceOverrides{},
 		DefaultSerializer:       defaultSerializer,
-		SerializerOverrides:     map[unversioned.GroupResource]runtime.NegotiatedSerializer{},
 		ResourceEncodingConfig:  resourceEncodingConfig,
-		CohabitatingResources:   map[unversioned.GroupResource][]unversioned.GroupResource{},
 		APIResourceConfigSource: resourceConfig,
 
 		newEtcdFn: newEtcd,
 	}
 }
 
 func (s *DefaultStorageFactory) SetEtcdLocation(groupResource unversioned.GroupResource, location []string) {
-	s.EtcdLocationOverrides[groupResource] = location
+	overrides, exists := s.GroupResourceOverrides[groupResource]
+	if !exists {
```

Comment:
Due to go's defaulting system and the way you structured the map, you can omit every one of these if statements, and the code will work the same.